core-processor/src/ext.rs

```diff
@@ -319,6 +319,134 @@ impl GrowHandler for LazyGrowHandler {
     }
 }
 
+enum Change {
+    MarkUsed(ReservationId),
+}
+
+/// Used to atomically update `Ext` which prevents some errors
+/// when data was updated but operation failed.
+///
+/// Copies some counters into itself and performs operations on them and
+/// incrementally builds list of changes.
+///
+/// Changes are applied after operation is completed
+struct ExtChanges {
```

Comment:
just to consider: what do you think about changes just own counters and keeps amount to be burned/spent and on each try to charge something it sums them from underlying store with burn/charge argument and call something like `counter.can_charge()`?
that will allow us not to mutate, clone and replace anything?

Reply:
I've initially used this approach but it resulted in not so good looking code, I could try to make it look readable again and push the changes

Reply:
Let's give it another try in different pr and focus to apply these changes in any manner

Reply:
Feel free to share some thoughts in issue not to lose context